pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
 
     Parameters
     ----------
-    filename: str
-        A relative or absolute file path.
+    fbuf: file-like buffer
+        Buffer of a BSRN station-to-archive data file
+    logical_records: str or list, default: ['0100']
+        List of the logical records (LR) to parse. Options are: '0100', '0300',
+        and '0500'.
+
+    Returns
+    -------
+    data: DataFrame
+        timeseries data from the BSRN archive, see
+        :func:`pvlib.iotools.read_bsrn` for fields.
+    metadata: dict
+        Dictionary containing metadata (primarily from LR0004).
+
+    See Also
+    --------
+    pvlib.iotools.read_bsrn, pvlib.iotools.get_bsrn
+
+    """
+    # Parse metadata
+    fbuf.readline()  # first line should be *U0001, so read it and discard
+    date_line = fbuf.readline()  # second line contains important metadata
+    start_date = pd.Timestamp(year=int(date_line[7:11]),
+                              month=int(date_line[3:6]), day=1,
+                              tz='UTC')  # BSRN timestamps are UTC
+
+    metadata = {}  # Initilize dictionary containing metadata
+    metadata['start date'] = start_date
+    metadata['station identification number'] = int(date_line[:3])
+    metadata['version of data'] = int(date_line.split()[-1])
+    for line in fbuf:
+        if line[2:6] == '0004':  # stop once LR0004 has been reached
+            break
+        elif line == '':
+            raise ValueError('Mandatory record LR0004 not found.')
+    metadata['date when station description changed'] = fbuf.readline().strip()
+    metadata['surface type'] = int(fbuf.readline(3))
+    metadata['topography type'] = int(fbuf.readline())
+    metadata['address'] = fbuf.readline().strip().strip()
+    metadata['telephone no. of station'] = fbuf.readline(20).strip()
+    metadata['FAX no. of station'] = fbuf.readline().strip()
+    metadata['TCP/IP no. of station'] = fbuf.readline(15).strip()
+    metadata['e-mail address of station'] = fbuf.readline().strip()
+    metadata['latitude'] = float(fbuf.readline(8))
+    metadata['longitude'] = float(fbuf.readline(8))
+    metadata['altitude'] = int(fbuf.readline(5))
+    metadata['identification of "SYNOP" station'] = fbuf.readline().strip()
+    metadata['date when horizon changed'] = fbuf.readline().strip()
+    # Pass last section of LR0004 containing the horizon elevation data
+    horizon = []  # list for raw horizon elevation data
+    while True:
+        line = fbuf.readline()
+        if ('*' in line) | (line == ''):
+            break
+        else:
+            horizon += [int(i) for i in line.split()]
+    horizon = pd.Series(horizon[1::2], horizon[::2], name='horizon_elevation',
+                        dtype=int).drop(-1, errors='ignore').sort_index()
+    horizon.index.name = 'azimuth'
+    metadata['horizon'] = horizon
+
+    # Read file and store the starting line number and number of lines for
+    # each logical record (LR)
+    fbuf.seek(0)  # reset buffer to start of file

```

Comment:
What's the reason for `seek`ing back to the beginning here?  I think the data records will always be after `LR0004`, so is there any problem with just mapping the sections that appear after `LR0004`?

Reply:
If the seek is remove here, then we need to manually add the line number offset (where the metadata section ends) to `enumerate` using *start* as otherwise, the line numbering will be incorrect in the dictionary.

As the metadata section and the earlier sections have a variable number of lines this will probably be rather complex to keep track of.